### PR TITLE
OpenIDConnect v1 implementation

### PR DESCRIPTION
**Pull request by jthann**
OpenIDConnect integration implementation https://github.com/megaease/easegress/issues/827

### CODE REVIEW

File: pkg/filters/oidcadaptor/oidcadaptor.go [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+}
+
+func (o *Filter) Status() interface{} {
+	return nil
+}
+
+func (o *Filter) Close() {
+}
+
+func (o *Filter) initDiscoveryOidcConf() {
```
Comment (localvar) [inline pkg/filters/oidcadaptor/oidcadaptor.go:254]: per Go naming convention, and there are other cases.

```suggestion
func (o *Filter) initDiscoveryOIDCConf() {
```

File: pkg/filters/oidcadaptor/oidcadaptor.go [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+			return true
+		}
+		return false
+	}
+	if len(spec.CookieName) != 0 && cookieExists() {
+		return ""
+	}
+	if bearerTokenExists() {
+		return ""
+	}
```
Comment (localvar) [inline pkg/filters/oidcadaptor/oidcadaptor.go:203]: propose not using the closure functions, they make code hard to read in this case and can be avoid.

File: pkg/filters/oidcadaptor/oidcadaptor.go [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+	if req.Path() == o.spec.redirectPath {
+		return o.handleOidcCallback(ctx)
+	}
+	//https://openid.net/specs/openid-connect-core-1_0.html#AuthorizationEndpoint
+	var authUrlB strings.Builder
```
Comment (localvar) [inline pkg/filters/oidcadaptor/oidcadaptor.go:208]: the name is not good, it took me some time to find where `authUrlA` is.

File: doc/reference/filters.md [hunk 4]
```diff
@@ -947,6 +950,76 @@ filters:
...
+
+
+### Results
+
+| Value           | Description                                                                                     |
+|-----------------|-------------------------------------------------------------------------------------------------|
+| X-Userinfo      | The End-User basic profile map json encoded by standard base64 encoding carried in HTTP header. |
+| X-Origin-ReqURL | The End-User origin request url before OpenID Connect or OAuth2.0 flow carried in HTTP header.  |
+| X-Id-Token      | The idToken returned by OpenID Connect flow carried in HTTP header.                             |
+| X-Access-Token  | The accessToken returned by OpenId Connect or OAuth2.0 flow carried in HTTP header.             |
```
Comment (localvar) [inline doc/reference/filters.md:1020]: the result listed here do not match the code.

File: pkg/filters/oidcadaptor/oidcadaptor.go [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+
+	//Following are user custom defined
+	jwksRefreshInterval  string
+	setAccessTokenHeader bool
+	setIDTokenHeader     bool
+	setUserInfoHeader    bool
+
+	redirectPath string
+	oidcConfig   *oidcConfig
+	jwks         *keyfunc.JWKS
```
Comment (localvar) [inline pkg/filters/oidcadaptor/oidcadaptor.go:104]: `Spec` is mainly for static configurations, these fields can be defined in the Filter.

**Review by suchen-sci — APPROVED** [review]

**Pull request merged**